### PR TITLE
Rtab setup

### PR DESCRIPTION
**Pull request by collinsc**
# Goals
- implement some basic infrastructure.
- get rtabmap up and running.
- create basic testing infrastructure to allow for sensor detection and movement in gazebo

# Works Completed
Speed was a priority here, so I adapted some old work from gazebo and implemented a robot model, called proto, proto is described in the tfr_description, and tfr_control packages. All related files are given the proto prefix.

I also created a test course which has a hallway leading into a room with a few obstaclesin the gazebo simulation, and saved it in the tfr_gazebo worlds folder.

I created the tfr_visualization package which is where we can store various different rviz visualization settings for our data.

I also created a testing configuration in the tfr_launch package called rtab_test.launch. This will set up the robot in the world and start rtab map, and an rviz view showing the pointcloud and generated map.

# Results
So what I found here is that I don't think the plugins for stereo cameras in gazebo are really going to do the thing we want. They are spotty, hard to configure, and require so much effort any time saved by not having to run on hardware is paid back in full. I think trying to make that happen for the navigation stack will be folly.

So I propose changing strategies and moving into hardware testing early. However I accomplished what I set out to do, and there is enough progress made to check my work in.

### CODE REVIEW

File: src/tfr_control/src/ProtoDriverStation.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    window.clear(sf::Color::Black);
+
+    window.display();
+    ros::spinOnce();
+    window.setFramerateLimit(30);
+  }
+
+  return 0;
+}
```
Comment (RyanBerge) [inline src/tfr_control/src/ProtoDriverStation.cpp]: Not that it matters too much, but let's keep our .cpp files all formatted the same way; that is, 4-space tabs and aligned braces.  I'm not sure how I feel about `case` statements wrapped in braces, but it's probably fine.

File: src/tfr_control/src/ProtoDriverStation.cpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
+/**
```
Comment (AdamuKaapan) [inline src/tfr_control/src/ProtoDriverStation.cpp]: File name should be all lowercase: should be proto_driver_station.cpp

File: src/tfr_description/xacro/proto_drivebase_controller.xacro [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      <rightFrontJoint>wheel_fr_joint</rightFrontJoint>
+      <leftRearJoint>wheel_rl_joint</leftRearJoint>
+      <rightRearJoint>wheel_rr_joint</rightRearJoint>
+      <wheelSeparation>${base_width}</wheelSeparation>
+      <wheelDiameter>${wheel_radius*2}</wheelDiameter>
+      <robotBaseFrame>god</robotBaseFrame>
+      <torque>20</torque>
+      <topicName>cmd_vel</topicName>
+      <commandTopic>cmd_vel</commandTopic>
+      <!-- we are simulating the output of sensor fusion right here -->
```
Comment (AdamuKaapan) [inline src/tfr_description/xacro/proto_drivebase_controller.xacro:40]: Is this actually producing fused odometry? Or are you just saying that, in the absence of other sensors to do fusion, we're just using this as the fused odometry even though it's only one source?

Comment (AdamuKaapan) [reply src/tfr_description/xacro/proto_drivebase_controller.xacro:40]: (if that's the case, no worries, I'm just clarifying)

File: src/tfr_description/xacro/proto_sensors.xacro [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        <noise>
+          <type>gaussian</type>
+          <mean>0.0</mean>
+          <stddev>0.0007</stddev>
+        </noise>
+      </camera>
+  </xacro:macro>
+  
+  <gazebo reference="camera_link">
+      <sensor type="multicamera" name="stereo_camera">
```
Comment (AdamuKaapan) [inline src/tfr_description/xacro/proto_sensors.xacro]: Indentation issue, there should be one less tab here before <sensor>

File: src/tfr_description/xacro/proto_drivebase_controller.xacro [hunk 5]
```diff
@@ -0,0 +1,49 @@
+<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="proto">
+  <!--xacro dectiptio of the prototype skid steer controller-->
```
Comment (AdamuKaapan) [inline src/tfr_description/xacro/proto_drivebase_controller.xacro]: Typo: I'm not sure, but is this supposed to be "depiction"?

File: src/tfr_launch/package.xml [hunk 6]
```diff
@@ -0,0 +1,60 @@
+<?xml version="1.0"?>
+<package format="2">
+  <name>tfr_launch</name>
+  <version>0.0.0</version>
+  <description>contains various system level starting configurations to accomplish various tasks</description>
```
Comment (AdamuKaapan) [inline src/tfr_launch/package.xml]: Capitalize "contains"?

File: src/tfr_visualization/package.xml [hunk 7]
```diff
@@ -0,0 +1,60 @@
+<?xml version="1.0"?>
+<package format="2">
+  <name>tfr_visualization</name>
+  <version>0.0.0</version>
+  <description>The tfr_visualization package</description>
```
Comment (AdamuKaapan) [inline src/tfr_visualization/package.xml]: More descriptive description would be nice so we know exactly what goes in this package? Is it just rviz scripts? Any nodes that involve visualization even if they're custom written?

**Review by AdamuKaapan — CHANGES_REQUESTED** [review]
Left various comments regarding a couple questions and a couple typos.

File: src/tfr_control/src/ProtoDriverStation.cpp (hunk 2 above)
Comment (collinsc) [reply src/tfr_control/src/ProtoDriverStation.cpp]: Good call, thanks for catching that we really should incorporate this style guide in future reviews: http://wiki.ros.org/CppStyleGuide

File: src/tfr_description/xacro/proto_drivebase_controller.xacro (hunk 5 above)
Comment (collinsc) [reply src/tfr_description/xacro/proto_drivebase_controller.xacro]: depiction, I'll fix it

File: src/tfr_control/src/ProtoDriverStation.cpp (hunk 2 above)
Comment (AdamuKaapan) [reply src/tfr_control/src/ProtoDriverStation.cpp]: It's already in the style guide, yup. Well, at least the naming is anyways

File: src/tfr_description/xacro/proto_drivebase_controller.xacro (hunk 3 above)
Comment (collinsc) [reply src/tfr_description/xacro/proto_drivebase_controller.xacro:40]: The reason I did that was to make sure that rtab map was subscribed to the correct topic, the idea was that we would be able to swap the gazebo simulation with the real thing with no issues.

Comment (AdamuKaapan) [reply src/tfr_description/xacro/proto_drivebase_controller.xacro:40]: Got it, I figured it was something like that.

File: src/tfr_visualization/package.xml (hunk 7 above)
Comment (collinsc) [reply src/tfr_visualization/package.xml]: I just forgot to add it, I'll elaborate.

**Comment by collinsc** [comment]
@AdamuKaapan I implemented the recommended changes. Thanks for catching them.

**Review by AdamuKaapan — DISMISSED** [review]

**Comment by collinsc** [comment]
@AdamuKaapan @RyanBerge I had to fix a merge conflict with dependencies, it was one line so it needs review again

**Review by AdamuKaapan — DISMISSED** [review]

**Review by RyanBerge — CHANGES_REQUESTED** [review]
I added a comment on `proto_driver_station.cpp` for changing the format to match our other source files.

File: src/tfr_control/src/ProtoDriverStation.cpp (hunk 1 above)
Comment (collinsc) [reply src/tfr_control/src/ProtoDriverStation.cpp]: I am following the ros c++ style guide. There is no guideline established for this in our coding standard, so I would prefer ros, which is 2 spaces for indents, with softtabs.

Comment (collinsc) [reply src/tfr_control/src/ProtoDriverStation.cpp]: lets' talk at the meeting

**Review by AdamuKaapan — APPROVED** [review]

**Review by RyanBerge — APPROVED** [review]

**Pull request merged**